frame/vesting/src/lib.rs

```diff
@@ -302,72 +409,273 @@ pub mod pallet {
+			// merged schedule.
+			return Some(schedule2);
+		} else if schedule2_ending_block <= now_as_balance {
+			return Some(schedule1);
+		}
+
+		let locked = schedule1
+			.locked_at::<T::BlockNumberToBalance>(now)
+			.saturating_add(schedule2.locked_at::<T::BlockNumberToBalance>(now));
+		// This shouldn't happen because we know at least one ending block is greater than now.
+		if locked.is_zero() {
+			return None;
+		}
+
+		let ending_block = schedule1_ending_block.max(schedule2_ending_block);
+		let starting_block = now.max(schedule1.starting_block()).max(schedule2.starting_block());
+		let duration =
+			ending_block.saturating_sub(T::BlockNumberToBalance::convert(starting_block));
+		let per_block = if duration.is_zero() {
+			// The logic of `ending_block` guarantees that each schedule ends at least a block
+			// after it starts and since we take the max starting and ending_block we should never
+			// get here
+			locked
+		} else if duration > locked {
+			// This would mean we have a per_block of less than 1, which should not be not possible
+			// because when we create the new schedule is at most the same duration as the longest,
+			// but never greater.
+			1u32.into()
+		} else {
+			locked.checked_div(&duration)?
+		};
+
+		// At this point inputs have been validated, so this should always be `Some`.
+		VestingInfo::try_new::<T>(locked, per_block, starting_block).ok()
+	}
+
+	// Execute a vested transfer from `source` to `target` with the given `schedule`.
+	fn do_vested_transfer(
+		source: <T::Lookup as StaticLookup>::Source,
+		target: <T::Lookup as StaticLookup>::Source,
+		schedule: VestingInfo<BalanceOf<T>, T::BlockNumber>,
+	) -> DispatchResult {
+		VestingInfo::validate_params::<T>(
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		)?;
+
+		let target = T::Lookup::lookup(target)?;
+		let source = T::Lookup::lookup(source)?;
+		if let Some(len) = Vesting::<T>::decode_len(&target) {
+			ensure!(
+				len < T::MaxVestingSchedules::get() as usize,
+				Error::<T>::AtMaxVestingSchedules
+			);
+		}
+
+		T::Currency::transfer(
+			&source,
+			&target,
+			schedule.locked(),
+			ExistenceRequirement::AllowDeath,
+		)?;
+
+		// We can't let this fail because the currency transfer has already happened
+		Self::add_vesting_schedule(
+			&target,
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		)
+		.expect("schedule inputs and vec bounds have been validated. q.e.d.");
+
+		Ok(())
+	}
+
 	/// (Re)set or remove the pallet's currency lock on `who`'s account in accordance with their
-	/// current unvested amount.
-	fn update_lock(who: T::AccountId) -> DispatchResult {
-		let vesting = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+	/// current unvested amount and prune any vesting schedules that have completed.
+	///
+	/// NOTE: This will update the users lock, but will not read/write the `Vesting` storage item.
```

Comment:
> but will not read/write the `Vesting` storage item

I don't think this is the ideal api, but I wanted to allow for only reading and writing to the vesting vec once for modification, and passing in through here seemed like the best compromise.